### DISCUSSION TITLE
Idea: blurhash and humbhash auto generated for images

### DISCUSSION THREAD

**Discussion by austriker27** [johannschopplich/unlazy#11] [Ideas]
From what I can tell, the `thumbhash` (and `blurhash`) must be provided a string. It would be a great feature if this was generated automatically either server or client side. (Similar to what Gatsby Image did years ago)

Thanks!

**Comment by johannschopplich** [comment]
Good idea in general. Although, unlazy is intended as a client-first library at the moment. Only the Nuxt module has meta-framework context and would be able to convert an image into a hash. 

A better fit is probably to find a way to include unlazy into the Nuxt Image module, respectively the to-go solution in the Next.js world (which I'm not familiar with).

**Comment by chrisspiegl** [comment]
I have achieved something similar to this by wraping the `UnLazyImage` component in my own component and using `@Nuxt/Image` to do the image rendering.

In this particular code example, I am not having it render the Thumbhash but instead using the `@Nuxt/Image` component to replace the domain names and creating the `srcset` for multiple sizes while providing resizing options through `IPX`.

In a similar way, you may be able to integrate the `Thumbhash` server side generation by doing something like the code provided below which is my server side function code to generate a Thumbhash from a uploaded image in this case.

### Custom Image Component to do stuff between NiyamaImage & UnLazyImage

```vue
<script setup lang="ts">
import type { ImageModifiers, ImageSizes } from '@nuxt/image';
import { consola } from 'consola';
import type { BlurHashObject } from '~~/types/database';

const { src, sizes, preset, densities, provider, modifiers, thumbhash, blurhash, preload, lazyLoad } = withDefaults(defineProps<{
  src?: string
  preload?: boolean
  lazyLoad?: boolean
  preset?: string
  densities?: string
  provider?: string
  modifiers?: Partial<ImageModifiers>
  sizes?: string
  thumbhash?: string
  blurhash?: BlurHashObject
  ssr?: boolean
}>(), {
  ssr: true,
})

const logger = consola.withTag('Component/NiyamaImage')
// logger.debug('🛠️ Setup')

const img = useImage()
const presets = img.options.presets

const _srcset = computedEager(() => {
  const _preset = preset ? presets[preset] : undefined
  const _sizes = (!sizes && _preset) ? _preset?.sizes : sizes
  const _modifiers = (!modifiers && _preset) ? _preset.modifiers : modifiers
  if (!src) {
    return {
      srcset: '',
      sizes: '',
      src: '',
    } as ImageSizes
  }
  return img.getSizes(src, {
    sizes: _sizes,
    modifiers: _modifiers,
    preset,
    densities,
    provider,
  }, _sizes)
})

const _blurhash = computedEager(() => ({
  hash: blurhash?.hash || '',
  size: blurhash?.width || 0,
  ratio: blurhashRatio(blurhash) || 1,
}))
</script>

<template>
  <UnLazyImage
    :key="_srcset.src || src"
    :src="_srcset.src || src"
    :src-set="_srcset.srcset"
    :sizes="_srcset.sizes"
    :thumbhash="thumbhash"
    :blurhash="_blurhash.hash"
    :placeholder-size="_blurhash.size"
    :placeholder-ratio="_blurhash.ratio"
    :selectable="false"
    :draggable="false"
    :preload="preload || false"
    :loading="lazyLoad ? 'lazy' : ''"
    :ssr="ssr"
  />
</template>

```


### Custom Function to Generate Thumbhash

This code would obviously have to be wrapped to only execute on the server.

```js
import { Buffer } from 'node:buffer'
import { createCanvas, loadImage } from '@napi-rs/canvas'
import { rgbaToThumbHash } from 'thumbhash'

export async function generateThumbhash(bufferBody, maxSize = 100) {
  const image = await loadImage(bufferBody)
  const { width, height } = image
  const scale = Math.min(maxSize / width, maxSize / height)
  const resizedWidth = Math.floor(width * scale)
  const resizedHeight = Math.floor(height * scale)

  const canvas = createCanvas(resizedWidth, resizedHeight)
  const context = canvas.getContext('2d')
  context.drawImage(image, 0, 0, resizedWidth, resizedHeight)

  const imageData = context.getImageData(0, 0, resizedWidth, resizedHeight)
  const rgba = new Uint8Array(imageData.data.buffer)
  const binaryThumbHash = rgbaToThumbHash(imageData.width, imageData.height, rgba)

  return {
    hash: Buffer.from(binaryThumbHash).toString('base64'),
    width: imageData.width,
    height: imageData.height,
  }
}
```

**Reply by johannschopplich** [reply]
That looks promising! We could implement that for the Nuxt UnLazy integration (is this possible without Nuxt Image and directly using IPX)?, as it's the only SSR framework integration for now. PR welcome! ⚡

**Reply by chrisspiegl** [reply]
[IPX](https://www.npmjs.com/package/ipx) is available individually and can be integrated by itself.

However, there is a bunch of trickery going on in the `@nuxt/image` library which makes it a module for Nuxt and thus also possible to run the image resizer directly in the Nuxt Server API.

However, the code that renders the Thumbhash on the server has nothing todo with `@nuxt/image` and could be implemented completely standalone.

Also, I do have a second function which generates Blurhash images. I just don't use that myself any more since we decided to go with Thumbhash from here on out.

**Reply by johannschopplich** [reply]
Indeed, but using IPX standalone is quite tricky. At least in my experience. 😄 We also could conditionally use Nuxt Image if available.
The ThumbHash and BlurHash generation logic would be a great addition to the unlazy core. 👍

**Reply by chrisspiegl** [reply]
That's true with IPX. I think there is a great in between spot though:

- Implement IPX for
	- automatic url generating for the different cloud image resizing providers with presets.
- Relying on `@nuxt/image`
	- for the "in built IPX server" nuxt module.

This would yield the possibility to use IPX for all the cloud providers and if one really wants to use the IPX resizing server on their own function, then they can use `@nuxt/image` module.

I personally for example am using the `@nuxt/image` in my Nuxt app only for the url generating stuff. But then I run my own `IPX-Server-Function` which I use for multiple projects (with a bigger RAM etc).

### My Implementations (pulled together from different gist and github resources):

These two functions render the `blurhash` and `thumbhash` from a image buffer. As far as I know these are pretty performant implementations using the the `@napi-rs/canvas`.

```js
import { Buffer } from 'node:buffer'
import { createCanvas, loadImage } from '@napi-rs/canvas'
import { rgbaToThumbHash } from 'thumbhash'
import { encode as rgbaToBlurhash } from 'blurhash'

export async function generateBlurhash(bufferBody, maxSize = 100) {
  const image = await loadImage(bufferBody)
  const { width, height } = image
  const scale = Math.min(maxSize / width, maxSize / height)
  const resizedWidth = Math.floor(width * scale)
  const resizedHeight = Math.floor(height * scale)

  const canvas = createCanvas(resizedWidth, resizedHeight)
  const context = canvas.getContext('2d')
  context.drawImage(image, 0, 0, resizedWidth, resizedHeight)

  const imageData = context.getImageData(0, 0, resizedWidth, resizedHeight)
  const binaryThumbHash = rgbaToBlurhash(imageData.data, imageData.width, imageData.height, 4, 4)

  return {
    hash: binaryThumbHash,
    width: imageData.width,
    height: imageData.height,
  }
}

export async function generateThumbhash(bufferBody, maxSize = 100) {
  const image = await loadImage(bufferBody)
  const { width, height } = image
  const scale = Math.min(maxSize / width, maxSize / height)
  const resizedWidth = Math.floor(width * scale)
  const resizedHeight = Math.floor(height * scale)

  const canvas = createCanvas(resizedWidth, resizedHeight)
  const context = canvas.getContext('2d')
  context.drawImage(image, 0, 0, resizedWidth, resizedHeight)

  const imageData = context.getImageData(0, 0, resizedWidth, resizedHeight)
  const rgba = new Uint8Array(imageData.data.buffer)
  const binaryThumbHash = rgbaToThumbHash(imageData.width, imageData.height, rgba)

  return {
    hash: Buffer.from(binaryThumbHash).toString('base64'),
    width: imageData.width,
    height: imageData.height,
  }
}
```

**Reply by luksak** [reply]
Hi @chrisspiegl ! Thank you for the code examples! I've been struggling to get this working in my nuxt codebase. How would you call those functions? In my case I'm using `@nuxt/content`. Could you provide an example of how you implemented this in a nuxt application? Thanks a lot for your work :heart:

**Reply by luksak** [reply]
Ok, i realized that I was trying to solve this using the wrong approach. In case of a static site built with `@nuxt/content`, the hashes need to be generated and stored somewhere during prerendering. I'm currently looking into how to approach this. In case anyone has an approach to this, I'd be thankful in case you could share some other approaches.

**Reply by luksak** [reply]
Oh and here is where my tests ended with my other approach: It is very hacky by using state to store the hash to fix it during hydration. But it works on the initial load. Rendering in the client is broken.


`ResponsiveImage.vue`:
```vue 
<template>
  <UnLazyImage
    :src
    :src-set="srcset"
    :alt="alt"
    loading="lazy"
    auto-sizes
    :thumbhash="thumbhash" />
</template>

<script setup lang="ts">
interface Props {
  src: string
  alt: string
  sizes?: string
  loading?: string
  aspectRatio?: string | boolean
  parentFit?: string | boolean
}

const props = withDefaults(defineProps<Props>(), {
  sizes: '100vw',
  loading: 'lazy',
  aspectRatio: false,
  parentFit: false,
})

const srcset = computed<string>(() => {
  const img = useImage()
  return img.getSizes(props.src, {
    sizes:
      'xs:100vw sm:100vw md:100vw lg:100vw xl:100vw xxl:100vw xxxl:100vw xxxxl:100vw xxxxxl:100vw',
    modifiers: {
      format: 'webp',
      quality: 60,
    },
  }).srcset
})

// Generate a unique key
const thumbhashKey = `thumbhash-${props.src}`

// Initialize state for thumbhash
const thumbhash = useState<string | null>(thumbhashKey, () => null)

// Generate thumbhash on server only
if (import.meta.server && !thumbhash.value) {
  const { generateThumbhash } = await import('~/server/utils/placeholder')
  thumbhash.value = await generateThumbhash(props.src)
}
</script>
```

`server/utils/placeholder.ts`
```ts
import { Buffer } from 'node:buffer'
import * as ThumbHash from 'thumbhash'
import sharp from 'sharp'

export async function generateThumbhash(path: string, maxSize = 100) {
  const image = sharp(`/home/luksak/Sites/nuxt-attribute-content/.playground/public${path}`).resize(100, 100, { fit: 'inside' })
  const { data, info } = await image.ensureAlpha().raw().toBuffer({ resolveWithObject: true })
  const binaryThumbHash = ThumbHash.rgbaToThumbHash(info.width, info.height, data)
  return Buffer.from(binaryThumbHash).toString('base64');
}
```

**Reply by chrisspiegl** [reply]
Looks like you found something of a solution. We do not generate our pages and use SSR instead. So that's why we mainly rely on the thumbhash build on image upload, stored in database, and then rendered via unlazy on the website.

**Reply by johannschopplich** [reply]
Just so you know, feel free to open a PR with your solutions if they offer a generic approach! 🙏

**Reply by luksak** [reply]
@chrisspiegl @johannschopplich  thank you for your feedback!

I realized that I need a solution for `@nuxt/content` specifically. For now I got a hacky but working version for my nuxt layer :balloon:  Here is the PR with the changes:

https://github.com/attributech/nuxt-attribute-content/pull/43

Finalizing that tomorrow. Let's see if we can come up with something generic. I'd be happy to contribute that if I can.

**Reply by luksak** [reply]
Ok, I've got a [working version released](https://github.com/attributech/nuxt-attribute-content/releases/tag/v0.1.6). The missing part was a nuxt content transformer. Here are the relevant parts:

[nuxt.config.ts](https://github.com/attributech/nuxt-attribute-content/blob/5e1740f179a72257a7d7f8f22a76852e01b7be44/nuxt.config.ts#L20C1-L26C5):
```ts 
export default defineNuxtConfig({
  content: {
    build: {
      transformers: [
        '#layers/nuxt-attribute-content/transformers/imageProcessing',
      ],
    },
  },
})
```

[transformers/imageProcessing.ts](https://github.com/attributech/nuxt-attribute-content/blob/main/transformers/imageProcessing.ts):
```ts 
import { defineTransformer, type TransformedContent } from '@nuxt/content'
import { processImage } from '../utils/imageProcessing'
import type { ImageItem } from '../types/ImageItem'

export default defineTransformer({
  name: 'imageProcessing',
  extensions: ['.yml'],
  transform: async function (content: TransformedContent) {
    await processContent(content)
    return content
  },
})

const processContent = async (data: Record<string, unknown>) => {
  for (const property in data) {
    let item = data[property]
    if (item instanceof Object) {
      item = await processContent(item as Record<string, unknown>)
    }
    if (item && item instanceof Object && 'src' in item) {
      item = await processImage(item as ImageItem)
    }
    data[property] = item
  }
  return data
}
```

[utils/imageProcessing.ts](https://github.com/attributech/nuxt-attribute-content/blob/main/utils/imageProcessing.ts):
```ts 
import { resolvePath } from '@nuxt/kit'
import { readFile } from 'node:fs/promises'
import { imageMeta, type ImageMeta } from 'image-meta'
import { createCanvas, loadImage } from '@napi-rs/canvas'
import { rgbaToThumbHash } from 'thumbhash'
import type { ImageItem } from '../types/ImageItem'

export async function processImage(imageItem: ImageItem): Promise<ImageItem> {
  const imageData = await readImage(imageItem.src)
  const metadata = getImageMetadata(imageData)
  const thumbhash = await generateThumbhash(imageData)
  imageItem.thumbhash = thumbhash
  imageItem.metadata = metadata
  return imageItem
}

export function getImageMetadata(imageData: Buffer<ArrayBufferLike>): ImageMeta {
  try {
    return imageMeta(imageData)
  }
  catch {
    throw new Error(`Failed to get image metadata.`)
  }
}

export async function readImage(path: string): Promise<Buffer<ArrayBufferLike>> {
  try {
    const filePath = await resolvePath(`public${path}`)
    return await readFile(filePath)
  }
  catch {
    throw new Error(`Failed to read image: ${path}`)
  }
}

export async function generateThumbhash(bufferBody: Buffer<ArrayBufferLike>, maxSize = 100) {
  try {
    const image = await loadImage(bufferBody)
    const { width, height } = image
    const scale = Math.min(maxSize / width, maxSize / height)
    const resizedWidth = Math.floor(width * scale)
    const resizedHeight = Math.floor(height * scale)

    const canvas = createCanvas(resizedWidth, resizedHeight)
    const context = canvas.getContext('2d')
    context.drawImage(image, 0, 0, resizedWidth, resizedHeight)

    const imageData = context.getImageData(0, 0, resizedWidth, resizedHeight)
    const rgba = new Uint8Array(imageData.data.buffer)
    const binaryThumbHash = rgbaToThumbHash(imageData.width, imageData.height, rgba)

    return Buffer.from(binaryThumbHash).toString('base64')
  }
  catch {
    throw new Error(`Failed to generate thumbhash.`)
  }
}
```

What do you think?

[The lookup of the images is very hacky](https://github.com/attributech/nuxt-attribute-content/blob/5e1740f179a72257a7d7f8f22a76852e01b7be44/transformers/imageProcessing.ts#L20C5-L22C6). It recursively looks for any object having a property `src`. If it finds that, it tries to parse the image. This works for my use case, but of course this isn't generic enough. I didn't have an idea how to make this more generic. One of the reasons being that it is a bit more complex to provide config at build time using app.config.ts.

I just had the thought that this could live in a separate nuxt module providing processing of images for `@nuxt/content`. The reason being that it can be useful to determine image dimensions without using unlazy. And then we could add some documentation on how to integrate that module with `@unlazy/nuxt`. 

I'd be happy to contribute this. Would someone be willing to look into the implementation details with me? I'm not that experienced with publishing nuxt modules.
